keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -104,14 +109,55 @@ class ApplicationService(
 
   fun pin(user: String, application: String, pin: EnvironmentArtifactPin) {
     val config = repository.getDeliveryConfigForApplication(application)
-    repository.pinEnvironment(config, pin.copy(pinnedBy = user))
-    // TODO: publish ArtifactPinnedEvent
+    val time = repository.pinEnvironment(config, pin.copy(pinnedBy = user))
```

Comment:
I might be able to refactor notifications when there are common things as adding more notifications.